TesSense.py

```diff
@@ -33,142 +37,208 @@ def printerror(error, err) :                          # Error message with trunc
+
+                if chargedata['charging_state'] == "Disconnected": # Loop w/o msgs until connected
+                    if not pshown :
+                        printmsg(" Please plug the vehicle in...\n\n")
+                        pshown = True
+
+                    # Set 0 watts to plug
+                    mutable_plug.data_source.power = 0
+                    await asyncio.sleep(60)
+                    continue
+                else : pshown = False
+
+                if chargedata['battery_level'] >= chargedata['charge_limit_soc'] : # Loop when full
+                    # Set 0 watts to plug
+                    mutable_plug.data_source.power = 0
+                    print ("Full Battery!")
+                    await asyncio.sleep(60)
+                    continue
+
+                if chargedata['fast_charger_present']:    # Loop while DC Fast Charging
+                    printmsg("Supercharging...")
+                    if charge != chargedata['battery_level'] :
+                        charge = chargedata['battery_level']
+                        print( "\nLevel:",
+                            chargedata['battery_level'], "%, Limit",
+                            chargedata['charge_limit_soc'], "%, Rate",
+                            chargedata['charger_power'], "kW, ",
+                            chargedata['charger_voltage'], "Volts, ",
+                            chargedata['fast_charger_type'],
+                            chargedata['minutes_to_full_charge'], "Minutes remaining\n" )
+
+                    # Set 0 watts to plug
+                    mutable_plug.data_source.power = 0
+                    await asyncio.sleep(60)
+                    continue
+
+                if chargedata['charging_state'] == "Charging": # Collect charging info
+                    volts =   chargedata['charger_voltage']
+                    rate =    chargedata['charge_current_request']
+                    maxrate = chargedata['charge_current_request_max']
+                    newrate = min( rate + int( power_diff / volts ), maxrate )
+                    if newrate == 1 or newrate == 2 : newrate = 0
+                    charging = True
+                else:
+                    charging = False
+
+
+            if charging :                                 # check if need to change rate or stop
+                if power_diff > 1 :
+                    print( "Charging at", rate, "amps, with", power_diff, "watts surplus" )
+                    if newrate > rate :
+                        print( "Increasing charging to", newrate, "amps" )
+                        try :
+                            vehicles[0].command( 'CHARGING_AMPS', charging_amps = newrate )
+                            if newrate < 5 : vehicles[0].command( 'CHARGING_AMPS', charging_amps = newrate )
+                            rate = newrate
+                            # Set plug power to new power?
+                            # mutable_plug.data_source.power = ...
```

Comment:
Somewhere in these if/else blocks you'd want to provide the "new" power reading to the plug. Alternatively, it may be smarter to simply do it at the end once it's been decided what to do (i.e. something like a `new_power` variable that you can use at the end).